x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/geo_threshold.ts

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { SearchResponse } from 'elasticsearch';
+import { executeEsQueryFactory, getShapesFilters, OTHER_CATEGORY } from './es_query_builder';
+import { AlertServices, AlertTypeState } from '../../../../alerts/server';
+import { ActionGroupId } from './alert_type';
+import { Logger } from '../../types';
+
+interface LatestEntityLocation {
+  location: string;
+  shapeLocationId: string;
+  entityName: string;
+  dateInShape: string;
+  docId: string;
+}
+
+// Flatten agg results and get latest locations for each entity
+function transformResults(
+  results: SearchResponse<unknown> | undefined,
+  dateField: string,
+  geoField: string
+): LatestEntityLocation[] {
+  if (!results) {
+    return [];
+  }
+
+  return (
+    _.chain(results)
+      .get('aggregations.shapes.buckets', {})
+      // @ts-expect-error
+      .flatMap((bucket: unknown, bucketKey: string) => {
+        const subBuckets = _.get(bucket, 'entitySplit.buckets', []);
+        return _.map(subBuckets, (subBucket) => ({
```

Comment:
I understand the aesthetic appeal of these functional constructs (chaining etc...) but they have a real impact on performance. 

Creating functions within loop-bodies (in this case, the one from `flatMap`) absolutely clobbers performance since it creates a new function object and lexical closure for each item in loop-body. This adds up real quickly, adding a real strain on the GC. In this case, you would create up to 65k temporary functions, just to process some results from vanilla JSON.

The alerting-framework is a paradigm shift in Kibana, where we are pushing a lot of processing to the server.

Given the size of the items in the array, the frequency with which this will run, and the system-context where it will run, I think this is a candidate to explicitly avoid the creation of a new closures inside a loop.

Reply:
Agreed this is definitely a candidate for future optimization. Unless there are any major objections, leaving this logic intact for this PR. It's a critical piece which has been more or less in place throughout the bulk of reviewer testing so would prefer not to modify it heavily. Future work could streamline this and or shift some of the work done here upstream to the original Elasticsearch query so that the shape of what's returned better aligns with what we'd prefer.